### ISSUE TITLE
[unified-signatures] Don't trigger on mergeable declaration with different parameter name

### ISSUE THREAD

**Issue by sindresorhus** [typescript-eslint/typescript-eslint#2386]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/unified-signatures": "error"
  }
}
```

```TS
declare const findUp: {
	(name: string | readonly string[], options?: findUp.Options): Promise<string | undefined>;

	(matcher: (directory: string) => (findUp.Match | Promise<findUp.Match>), options?: findUp.Options): Promise<string | undefined>;
};
```

**Expected Result**

I expected it not to be a violation as a different parameter name is a good indication that the overloads are not compatible.

**Actual Result**

```
  ✖   57:4  These overloads can be combined into one signature taking string | string[] | (directory: string) => findUp.Match.                            @typescript-eslint/unified-signatures
  ✖  134:3  These overloads can be combined into one signature taking string | string[] | (directory: string) => (findUp.Match | Promise<findUp.Match>).  @typescript-eslint/unified-signatures
```

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `3.9.0` |
| `@typescript-eslint/parser`        | `3.9.0` |
| `TypeScript`                       | `3.9.7` |
| `ESLint`                           | `7.6.0` |
| `node`                             | `10.22.0` |
| `npm`                              | `6.14.7` |

**Comment by bradzacher** [comment]
> This rule aims to keep the source code as maintainable as possible by reducing the amount of overloads.

If the signatures themselves are mergeable, then it seems like it aligns with the rule to merge them, regardless of parameter names.

I wouldn't hugely want to rely upon parameter names in this rule.

> different parameter name is a good indication that the overloads are not compatible

In some cases, maybe. In others - no.

```ts
declare const foo: {
  (argument: string);
  (argment: boolean);
  (arg: number);
}
```

Are all 3 signatures here unifiable or not? Probably not.. Could probably implement a heuristic like the [Levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance) to check for this. It's a lot of code to maintain for an rarer case / requirement though.

**Comment by sindresorhus** [comment]
> If the signatures themselves are mergeable, then it seems like it aligns with the rule to merge them, regardless of parameter names.

The overloads I showed are completely separate though, functionality-wise, and in the actual code, they have separate doc comments too.

Here's the actual code: https://github.com/sindresorhus/find-up/blob/e852e9cfd020844fac0ea2ff073c378cc302db4d/index.d.ts#L84-L135

> In some cases, maybe. In others - no.

That example is not realistic. A realistic example would be something like:

```ts
declare const foo: {
  (argument: string);
  (arguments: string[]);
}
```

> Could probably implement a heuristic like the Levenshtein distance to check for this.

That's too far, I think. But a simple pluralization check could make sense.

**Comment by bradzacher** [comment]
Happy to accept a PR for a default off option which tells the rule to ignore signatures if the param names are different.
